File: protocol/impls/csharp/Digest.cs
using System;
using System.Collections.Generic;
using System.Buffers.Binary;

namespace Pepenet;

/// <summary>
/// Canonical SHA-256 state digest (SPEC-conformance §4). All multi-byte integers
/// little-endian; signed two's-complement LE. Sorts are explicit and unsigned-bytewise;
/// Dictionary/HashSet enumeration never feeds the digest. Magic "SMv1". Names-only:
/// tables = names + commits + muts (no votes/decors/overflow).
/// </summary>
public static class Digest
{
    public static byte[] Compute(Fold f)
    {
        var buf = new Buf();
        buf.Ascii("SMv1");

        // ---- names: sorted ascending by raw name bytes ----
        var names = new List<NameRow>(f.Names.Values);
        names.Sort((a, b) => ByteArrayComparer.Instance.Compare(a.Name, b.Name));
        buf.U32((uint)names.Count);
        foreach (var r in names)
        {
            buf.U8((byte)r.Name.Length);
            buf.Bytes(r.Name);
            buf.Fixed(r.Owner, 20);
            buf.U8(r.St);
            buf.I64(r.LeaseExpiry);
            buf.Fixed(r.Seller, 20);
            buf.U8(r.SellerType);
            buf.U64(r.Price);
            buf.I64(r.OfferExpiry);
            buf.Fixed(r.Buyer, 20);
            buf.U64(r.BurnLeg);
            buf.U64(r.PayLeg);
            buf.I64(r.ReserveExpiry);
        }

        // ---- commits: sorted by (commitment[32], commit_height, tx_index) — total order ----
        var commits = new List<CommitRow>(f.Commits);
        commits.Sort((a, b) =>
        {
            int c = ByteArrayComparer.Instance.Compare(a.Commitment, b.Commitment);
            if (c != 0) return c;
            c = a.CommitHeight.CompareTo(b.CommitHeight);
            if (c != 0) return c;
            return a.TxIndex.CompareTo(b.TxIndex);
        });
        buf.U32((uint)commits.Count);
        foreach (var c in commits)
        {
            buf.Fixed(c.Commitment, 32);
            buf.I64(c.CommitHeight);
            buf.U32(c.TxIndex);
            buf.I64(c.CommitTime);
        }

        // ---- muts: sorted by owner bytes ----
        var muts = new List<(byte[] owner, long h)>();
        foreach (var kv in f.Muts)
            muts.Add((f.MutOwnerBytes[kv.Key], kv.Value));
        muts.Sort((a, b) => ByteArrayComparer.Instance.Compare(a.owner, b.owner));
        buf.U32((uint)muts.Count);
        foreach (var m in muts)
        {
            buf.Fixed(m.owner, 20);
            buf.I64(m.h);
        }

        return Hashing.Sha256(buf.ToSpan());
    }

    public static string ComputeHex(Fold f) => Hashing.Hex(Compute(f));

    private sealed class Buf
    {
        private byte[] _b = new byte[256];
        private int _len;

        private void Ensure(int n) { if (_len + n > _b.Length) Array.Resize(ref _b, Math.Max(_b.Length * 2, _len + n)); }
        public void U8(byte v) { Ensure(1); _b[_len++] = v; }
        public void Bytes(byte[] v) { Ensure(v.Length); Buffer.BlockCopy(v, 0, _b, _len, v.Length); _len += v.Length; }
        public void Fixed(byte[] v, int n)
        {
            Ensure(n);
            int copy = Math.Min(n, v.Length);
            Buffer.BlockCopy(v, 0, _b, _len, copy);
            for (int i = copy; i < n; i++) _b[_len + i] = 0;
            _len += n;
        }
        public void Ascii(string s) { Ensure(s.Length); foreach (char c in s) _b[_len++] = (byte)c; }
        public void U32(uint v) { Ensure(4); BinaryPrimitives.WriteUInt32LittleEndian(_b.AsSpan(_len, 4), v); _len += 4; }
        public void U64(ulong v) { Ensure(8); BinaryPrimitives.WriteUInt64LittleEndian(_b.AsSpan(_len, 8), v); _len += 8; }
        public void I64(long v) { Ensure(8); BinaryPrimitives.WriteInt64LittleEndian(_b.AsSpan(_len, 8), v); _len += 8; }
        public ReadOnlySpan<byte> ToSpan() => _b.AsSpan(0, _len);
    }
}

File: protocol/impls/go/digest.go
package main

import (
	"bytes"
	"crypto/sha256"
	"sort"
)

// stateDigest serializes the live state into the canonical byte layout pinned by
// SPEC-conformance.md §4 and returns its SHA-256. All multi-byte ints little-endian;
// signed values two's-complement LE. Magic "SMv1". Tables: names, commits, muts
// (names-only SM — votes/posts/decorations/overflow removed). Collections are
// emitted in the pinned sort order — NEVER Go map iteration order.
func (s *FoldState) stateDigest() [32]byte {
	var b bytes.Buffer
	b.WriteString("SMv1")

	// names — sorted ascending by raw name bytes
	nameKeys := make([]string, 0, len(s.names))
	for k := range s.names {
		nameKeys = append(nameKeys, k)
	}
	sort.Strings(nameKeys)
	wU32(&b, uint32(len(nameKeys)))
	for _, k := range nameKeys {
		r := s.names[k]
		b.WriteByte(byte(len(k)))
		b.WriteString(k)
		b.Write(r.owner[:])
		b.WriteByte(byte(r.st))
		wI64(&b, r.leaseExpiry)
		b.Write(r.seller[:])
		b.WriteByte(r.sellerType)
		wU64(&b, r.price)
		wI64(&b, r.offerExpiry)
		b.Write(r.buyer[:])
		wU64(&b, r.burnLeg)
		wU64(&b, r.payLeg)
		wI64(&b, r.reserveExpiry)
	}

	// commits — sorted by (commitment[32], commit_height, tx_index): total order
	commits := append([]CommitRow(nil), s.commits...)
	sort.Slice(commits, func(i, j int) bool {
		if c := bytes.Compare(commits[i].commitment[:], commits[j].commitment[:]); c != 0 {
			return c < 0
		}
		if commits[i].height != commits[j].height {
			return commits[i].height < commits[j].height
		}
		return commits[i].txIndex < commits[j].txIndex
	})
	wU32(&b, uint32(len(commits)))
	for _, c := range commits {
		b.Write(c.commitment[:])
		wI64(&b, c.height)
		wU32(&b, c.txIndex)
		wI64(&b, c.commitTime)
	}

	// muts — sorted by owner bytes
	owners := make([][20]byte, 0, len(s.muts))
	for o := range s.muts {
		owners = append(owners, o)
	}
	sort.Slice(owners, func(i, j int) bool {
		return bytes.Compare(owners[i][:], owners[j][:]) < 0
	})
	wU32(&b, uint32(len(owners)))
	for _, o := range owners {
		b.Write(o[:])
		wI64(&b, s.muts[o])
	}

	return sha256.Sum256(b.Bytes())
}

func wU32(b *bytes.Buffer, v uint32) {
	b.Write([]byte{byte(v), byte(v >> 8), byte(v >> 16), byte(v >> 24)})
}
func wU64(b *bytes.Buffer, v uint64) {
	var t [8]byte
	for i := 0; i < 8; i++ {
		t[i] = byte(v >> (8 * uint(i)))
	}
	b.Write(t[:])
}
func wI64(b *bytes.Buffer, v int64) { wU64(b, uint64(v)) }
